src/py2spack/package_providers.py

```diff
@@ -0,0 +1,316 @@
+"""Utils for downloading packages from PyPI and opening the pyproject.tomls."""
+
+from __future__ import annotations
+
+import abc
+import functools
+from typing import Any, Protocol
+
+import requests
+from packaging import version as vn
+
+from py2spack import utils
+
+
+KNOWN_ARCHIVE_FORMATS = [
+    ".tar",
+    ".tar.gz",
+    ".tar.bz2",
+    ".gz",
+    ".xz",
+    ".bz2",
+]
+
+
+def _parse_packaging_version(version: str) -> vn.Version | None:
+    """Parse packaging version."""
+    result = None
+    try:
+        v = vn.parse(version)
+        # do not support post releases of prereleases etc.
+        if not (v.pre and (v.post or v.dev or v.local)):
+            result = v
+    except vn.InvalidVersion:
+        pass
+
+    return result
+
+
+class PyProjectProviderQueryError(Exception):
+    """Error during querying of the PyProjectProvider."""
+
+
+class PyProjectProvider(Protocol):
+    """General provider interface for Python distribution packages."""
+
+    @abc.abstractmethod
+    def get_versions(self, name: str) -> list[vn.Version] | PyProjectProviderQueryError:
+        """Get available package versions.
+
+        Returns an error if no versions are found.
+        """
+
+    @abc.abstractmethod
+    def get_pyproject(
+        self, name: str, version: vn.Version
+    ) -> dict[Any, Any] | PyProjectProviderQueryError:
+        """Get the contents of the pyproject.toml file for the specified version."""
+
+    @abc.abstractmethod
+    def get_hash(
```

Comment:
I think it's worth to make the function name more explicit, since just `get_hash` is a bit too general.
```suggestion
    def get_sdist_hash(
```